src/dvc_data/index/index.py

```diff
@@ -222,6 +256,80 @@ def node_factory(_, key, children, *args):
 
         return self.traverse(node_factory, prefix=root_key)
 
+    def _filter(
+        self, filter_fn: Callable[[DataIndexKey], bool]
+    ) -> Iterator[DataIndexKey]:
+        """Iterate over keys matching the specified filter."""
+
+        class _FilterNode:
+            def __init__(self, key, children):
+                self.key = key
+                self.children = children
+
+            def __iter__(self):
+                if self.children:
+                    for child in self.children:
+                        if child is not None:
+                            yield from iter(child)
+                elif self.key:
+                    yield self.key
+
+        def node_factory(_, key, children, *args):
+            if not key or filter_fn(key):
+                return _FilterNode(key, children)
+
+        root = self.traverse(node_factory)
+        if root is not None:
+            yield from root
+
+    def filter(
+        self, filter_fn: Callable[[DataIndexKey], bool]
+    ) -> "DataIndexView":
+        """Return read-only filtered view of this index."""
```

Comment:
Right now the DVC-level repo `Index.filter()` returns an entire new `Index`, but for dvc-data I think what we want is a view into the existing trie (but with filtered keys)